src/charm.py

```diff
@@ -119,6 +152,56 @@ def _dashboards_relation_broken(self, _):
         self._stored.have_dashboards = False
         self._set_status()
 
+    def _on_stop(self, _):
+        """Ensure that nrpe exporter is removed and stopped."""
+        if service_running("nrpe-exporter"):
+            service_stop("nrpe-exporter")
+
+        files = ["/usr/local/bin/nrpe_exporter", "/etc/systemd/systemd/nrpe-exporter.service"]
+
+        for f in files:
+            if Path(f).exists():
+                Path(f).unlink()
+
+    def _nrpe_relation_joined(self, _):
+        self._stored.have_nrpe = True
+
+        # Make sure the exporter binary is present with a systemd service
+        if not Path("/usr/local/bin/nrpe_exporter").exists():
```

Comment:
I'd expect similar logic for upgrade charm, only on upgrade you'd probably need to compare hashes of `/usr/local/bin/nrpe_exporter` or analogous.

Reply:
Running as a machine charm, we can rely on it actually "sticking", since there's no pod destruction

Reply:
Can charm upgrade introduce a new version of the exporter to be used?

Reply:
For the time being, the future of who is going to maintain `nrpe_exporter` is up in the air, but the main repo hasn't really seen updates in years, so it's unlikely